Move gallery link controls to the block toolbar

akasunil opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Fixes: https://github.com/WordPress/gutenberg/issues/55022

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Required to move gallery block's "Link to" control from the inspector to a more familiar, and consistent, link control in the block toolbar.
 
## Screenshots or screencast <!-- if applicable -->

[Edit-Post-“Hello-world-”-‹-gutenberg-—-WordPress (1).webm](https://github.com/user-attachments/assets/2e967984-1a09-4990-a0ec-5c37e27889b0)

akasunil committed:
Move link dropdown to toolbar in gallery block

akasunil committed:
Update e2e test for gallery block's  setting

akasunil committed:
Fix e2e test for gallery block images link setting

akasunil committed:
Merge branch 'trunk' of personal.github.com:WordPress/gutenberg into update/gallery-block-image-link-control

akasunil committed:
Fix unit test

akasunil committed:
Synced with trunk and resolved conflicts

t-hamano reviewed (commented):
Thanks for the PR!

I did a quick test and found some layout issues:

- When the Gallery block is inside a Group block, the parent selector button is small in size.
- The link buttons seem to be slightly wider on both sides.

![image](https://github.com/user-attachments/assets/f253af04-564a-49a2-a9da-bc8318ec0abd)

I think the component to use here is `ToolbarDropdownMenu`.
If you search the `packages/block-libraty` directory for f`ToolbarDropdownMenu`, you will find several implementation examples.

Additionally, the text of the following messages has become longer and could be made clearer.

trunk:

![image](https://github.com/user-attachments/assets/9a7787ce-22da-4cc5-b94d-ce91db1d6443)

This PR:

![image](https://github.com/user-attachments/assets/7dc4267f-e80c-45dc-af5d-d2f51660a990)|

akasunil commented:
I appreciate your feedback.

> the text of the following messages has become longer and could be made clearer.

I updated the texts to reflect the suggestions made here on [issue](https://github.com/WordPress/gutenberg/issues/55022)

akasunil committed:
Synced with trunk

akasunil committed:
Added ToolbarDropdownMenu component instead of DropdownMenu

akasunil committed:
remove remove extra padding around link icon

akasunil committed:
Remove unused components

akasunil committed:
Revert "Synced with trunk"

This reverts commit 0c7a78b8569b436a73f1c8e43f2150a78f68601e.

akasunil committed:
Merge branch 'trunk' of personal.github.com:WordPress/gutenberg into update/gallery-block-image-link-control

akasunil commented:
Link icon issues are fixed now
<img width="1160" alt="image" src="https://github.com/user-attachments/assets/cd2e65d8-5b92-4d9a-8fc1-d34d771be5b8">

t-hamano reviewed (commented):
Thanks for the update! It's working as expected.

As a further improvement, I'm wondering if we can keep the notification text as suggested in [this comment](https://github.com/WordPress/gutenberg/pull/62762#pullrequestreview-2178003931).

Here are the changes I propose. What do you think?

<details><summary>diff</summary>

```diff
diff --git a/packages/block-library/src/gallery/edit.js b/packages/block-library/src/gallery/edit.js
index 1332c8d21b..7fda27ecdc 100644
--- a/packages/block-library/src/gallery/edit.js
+++ b/packages/block-library/src/gallery/edit.js
@@ -70,16 +70,19 @@ const linkOptions = [
                icon: customLink,
                label: __( 'Link images to attachment pages' ),
                value: LINK_DESTINATION_ATTACHMENT,
+               noticeText: __( 'Attachment Pages' ),
        },
        {
                icon: imageIcon,
                label: __( 'Link images to media files' ),
                value: LINK_DESTINATION_MEDIA,
+               noticeText: __( 'Media Files' ),
        },
        {
                icon: linkOff,
                label: _x( 'None', 'Media item link option' ),
                value: LINK_DESTINATION_NONE,
+               noticeText: __( 'None' ),
        },
 ];
 const ALLOWED_MEDIA_TYPES = [ 'image' ];
@@ -377,7 +380,7 @@ function GalleryEdit( props ) {
                        sprintf(
                                /* translators: %s: image size settings */
                                __( 'All gallery image links updated to: %s' ),
-                               linkToText.label
+                               linkToText.noticeText
                        ),
                        {
```

</details>

t-hamano commented:
Regarding the unit test failure, it seems that the block editor itself crashes on mobile, as [this action shows](https://github.com/WordPress/gutenberg/actions/runs/10019462053/job/27696050361?pr=62762).

We might be adding some code that is not compatible with mobile.

Unfortunately, I'm a Windows user and can't build the mobile app.

@geriux @dcalhoun, If you know why the unit tests are failing, I would appreciate it if you could let me know 🙇‍♂️ This PR moves the link control of the Gallery block from the block sidebar to the block toolbar.

akasunil committed:
Improve snackbar notices for gallery block link control select

t-hamano reviewed (approved):
LGTM! 

For unit test failures, let's wait for help from the mobile team 😅

paaljoachim commented:
It seems this PR will then do something like I mentioned here in number 1.
https://github.com/WordPress/gutenberg/pull/62906#issuecomment-2245014838

----

The Gallery needs a Link To option - Expand on click so that all the images inside a gallery will use Expand on click.
One part of this comment: https://github.com/WordPress/gutenberg/issues/37652#issuecomment-1819727874

Implement a way to apply the lightbox on setting to all the child images of the gallery. The gallery already does this with the linkto and resolution settings, so shouldn't be too much work

-----

akasunil commented:
@paaljoachim Do you mean that the `Expand on click` option that appears when we click the link icon in each gallery image should be included to the gallery link control?

<img width="911" alt="image" src="https://github.com/user-attachments/assets/ce9dbdf7-d20b-4dea-9d41-1279206ecc79">

paaljoachim commented:
@akasunil Yes

You got it. As the user would recognize what is seen in the Image block link Expand on click could then also be used in the Link to used in the Gallery block even though it is a bit hidden away.

akasunil committed:
Merge branch 'trunk' of personal.github.com:WordPress/gutenberg into update/gallery-block-image-link-control

akasunil commented:
We could do that. My recommendation was to address it in a different PR. However, we currently have a test case that is failing this PR, and we are awaiting comment from the mobile team. I will address the suggestion you made.

akasunil committed:
Add constant for lightbox option in link control for gallery block

akasunil committed:
Update getHrefAndDestination function to handle Link control value change for lightbox option

akasunil committed:
Add lightbox option in link control of gallery block

paaljoachim commented:
@akasunil Yes. It seems like a good idea to open a new PR addressing the things I mentioned. Then just adding a comment here to the other PR.

akasunil commented:
In fact, we may need to wait until this PR is merged before opening a new one due to the dependacy of this PR, as doing so will require more time to resolve conflicts following the merger. 

We can incorporate your suggested change into the current PR since it pertains to a specific capability that this PR addresses.

akasunil committed:
Add info text for lightbox option in link control for gallery block

akasunil committed:
Set info text for lightbox option in single line

akasunil commented:
Here is how it works

[Edit-Post-“Hello-world-”-‹-gutenberg-—-WordPress (1).webm](https://github.com/user-attachments/assets/2e967984-1a09-4990-a0ec-5c37e27889b0)

paaljoachim commented:
Hi @akasunil 

It looks like you added the link to the Gallery block toolbar.

Let's look at both toolbars.
The Gallery block has this toolbar.
![Screenshot 2024-07-24 at 18 52 16](https://github.com/user-attachments/assets/2d31329b-aecd-4694-8237-b481c773e48c)

The Gallery block has minor information in the toolbar and uses the sidebar settings more to control the Gallery.

The Image block has this toolbar.
![Screenshot 2024-07-24 at 18 52 30](https://github.com/user-attachments/assets/e58bdc3c-d1b6-4b0b-afdb-ad1badf741c7)

The Image block has a more focused toolbar approach.

Lets just make a minor adjustment right now and use the LINK To drop down feature in the right sidebar Gallery block settings to add the Expand on click. Adding adjustments to the toolbar is a bigger and more complex UI adjustment and is something that will be need to discussed in a broader discussion.

Lets keep to smaller adjustments. Keeping to this adjustment in relation to this comment from @glendaviesnz 
https://github.com/WordPress/gutenberg/issues/37652#issuecomment-1819727874

akasunil commented:
Actually, as stated in the PR description, the purpose of this PR is to relocate the `Link To` control from the sidebar to the toolbar as suggested in the issue https://github.com/WordPress/gutenberg/issues/55022. And i agree with that change, as image block have the `Link To` control in toolbar too. We should follow the same pattern for gallery block. 

If the same setting is in a different place for two different blocks, it will be confusing from the user point of view.

paaljoachim commented:
Ahhh gotcha. We are going full circle here.

paaljoachim commented:
@WordPress/gutenberg-design 
@jameskoster and @richtabor 

Any comments regarding the change of removing the Gallery block LINK TO dropdown in the sidebar settings and instead adding the same Image block toolbar link feature to the Gallery block toolbar?

t-hamano commented:
In my opinion, this PR is about moving existing functionality in the block sidebar to the block toolbar, and should avoid adding new functionality.

Looking at [this comment](https://github.com/WordPress/gutenberg/pull/62906#issuecomment-2245014838), it seems that a new option called "Expand on click" needs to be added to the Gallery block, and when enabled, it is expected that a UI will be displayed to control whether to show a chevron as an additional option.

If that is the case, would it be better not to move the link option to the block toolbar in the first place?

Either way, it may be better to hold off on this PR until the specs for #62906 are solidified.

akasunil commented:
> this PR is about moving existing functionality in the block sidebar to the block toolbar, and should avoid adding new functionality.

I acknowledge the concern and I have expressed my opinion regarding this on [comment](https://github.com/WordPress/gutenberg/pull/62762#issuecomment-2248367348)


> it seems that a new option called "Expand on click" needs to be added to the Gallery block, and when enabled, it is expected that a UI will be displayed to control whether to show a chevron as an additional option.

Since the "Expand on click" option is already present on the picture under the gallery block, it should be included there as well, since the gallery block already has all other `Link To` options identical to those on the image block. Coming to the slider capability in lightbox, i think that is something that needs discussion that i agree. But that is separate enhancement i believe.

By adding `Expand on click` option on gallery block, We are enabling the functionality that is already available on image block. Users can then enable the `Expand on click` option from any individual image block ( as it already available in trunk ) or can enable `Expand on click` from gallery block ( as this PR implement ), Both methods lead to the same outcome. Thus, I don't anticipate any problems.

akasunil committed:
Merge branch 'trunk' of personal.github.com:WordPress/gutenberg into update/gallery-block-image-link-control

t-hamano commented:
> By adding `Expand on click` option on gallery block, We are enabling the functionality that is already available on image block.

I understand this, of course.

However, we can't simply add the "Expand on click" option to the Gallery block, because we need to check whether the theme supports lightbox. And the current Image block has complex logic to decide whether to show this setting or not. Try searching for "lightbox" in the [packages/block-library/src/image/image.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/image.js) file.

Perhaps the Gallery block needs equivalent logic, and I think it will be necessary to thoroughly verify it in a separate PR to make it happen.

paaljoachim commented:
It really comes back to what I mentioned. It is a lot easier making smaller changes such as keeping the LINK To dropdown in the sidebar settings **for now** and just adding Expand into it. 

Adding something new even though it is consistent with the Image block is a lot more complex and really should be done at a later time.

I am circling back to this again. Simpler steps are easier to make compared to more complex steps. Breaking down complex steps into smaller simpler steps does help.

1. Keep the LINK To Gallery option in the sidebar and adding Expand into it. 
2. Discovering scenarios as @t-hamano Aki outlines that will impact the Expand option in the sidebar and addressing those. When these various concerns are adressed then...
3. One might be ready to make the next step as in moving the LINK To Gallery sidebar option to the Gallery Block toolbar. Again moving an option from the sidebar to the toolbar is a bigger step and one needs to figure out the smaller steps to get to it. That is also what I have been working on earlier adressing.

luisherranz commented:
Linking here a relevant conversation about the "Expand on click" option in galleries and pinging @joen.

- https://github.com/WordPress/gutenberg/issues/56310#issuecomment-2206080002

akasunil commented:
> And the current Image block has complex logic to decide whether to show this setting or not. Try searching for "lightbox" in the [packages/block-library/src/image/image.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/image.js) file.

Thank you for clarifying @t-hamano, I now understand your concern. In that case, should i revert the lightbox code?

paaljoachim commented:
I would suggest creating a new PR that only addresses the smallest step possible. Adding the "Expand on click" to the LINK To dropdown in the Gallery block sidebar settings. Addressing issues that @t-hamano mentioned in that PR. As there seems to be a few scenarios that will need to be addressed. Also going through the discussions at https://github.com/WordPress/gutenberg/issues/56310 which might impact the Expand option for the Gallery block.

madhusudhand reviewed (commented):
This works well in my testing.

Only Issue observed was the link icon is not set to active unlike the image, when a value is set.

<img width="518" alt="image" src="https://github.com/user-attachments/assets/ffa96ed6-f81d-4600-9276-e4b59d604b35">

Expectation: 

<img width="525" alt="image" src="https://github.com/user-attachments/assets/e642dc9b-552f-42a9-bf1f-bcc928425705">

t-hamano commented:
@madhusudhand I am hesitant to add the new "Expand on clock" option to the Gallery block in this PR because there are a lot of things that need to be taken into consideration. See [this comment](https://github.com/WordPress/gutenberg/pull/62762#issuecomment-2249532142).

madhusudhand commented:
@akasunil I tested the new "Expand on click" at gallery against [Gallery: Add lightbox support](https://github.com/WordPress/gutenberg/pull/62906) and it work perfectly fine. 

But it is best to bring the lightbox related changes into a new PR and keep the current PR specific to movement of the toolbar option, by doing so, it doesn't have to wait for #62906

madhusudhand commented:
> @madhusudhand I am hesitant to add the new "Expand on clock" option to the Gallery block in this PR because there are a lot of things that need to be taken into consideration. See https://github.com/WordPress/gutenberg/pull/62762#issuecomment-2249532142.

I agree.

akasunil commented:
Thank you @madhusudhand, I'll revert the change for `Expand on click` option, and create separate PR.

dcalhoun commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -594,6 +615,43 @@ function GalleryEdit( props ) {
+					label={ __( 'Link To' ) }
+				>
+					{ ( { onClose } ) => (
+						<MenuGroup className="blocks-gallery__link-to-control">
+							{ linkOptions.map( ( linkItem ) => {
+								const isOptionSelected =
+									linkTo === linkItem.value;
+								return (
+									<MenuItem
```
> @geriux @dcalhoun, If you know why the unit tests are failing, I would appreciate it if you could let me know 🙇‍♂️ This PR moves the link control of the Gallery block from the block sidebar to the block toolbar.

First, thank you for attempting to resolve the mobile unit test failures and for the ping. Much appreciated! 🙇🏻 

Both `MenuGroup` and `MenuItem` are unavailable for the native mobile editor currently. This results in the crash and unit test failure regarding the `Dropdown` component attempting to render/return an `undefined` value. 

We encountered this situation before and addressed it in https://github.com/WordPress/gutenberg/pull/35191 by bifurcating the code along the web and native platform line. It is not an ideal solution — see https://github.com/WordPress/gutenberg/pull/35191#issuecomment-930245860 — but, unfortunately, the lack of support for `MenuGroup` and `MenuItem` remains the current state. 

If we take a similar approach here, we might retain the previous `SelectControl` for native platforms. I opened https://github.com/WordPress/gutenberg/pull/63953 implementing this option, if we'd like to use this approach to move the proposed web changes forward quickly. 

Unfortunately, I do not believe the native mobile team will prioritize `MenuGroup` or `MenuItem` components in the near future.

akasunil replied:
@dcalhoun appreciate your response. Thank you for the details on the issue and PR too.

akasunil replied:
@t-hamano is it worth doing this way?

artemiomorales commented:
Hi all, after taking a look at this discussion and related ones in https://github.com/WordPress/gutenberg/pull/62906, I [posted a summary](https://github.com/WordPress/gutenberg/issues/56310#issuecomment-2252075254) with some potentially helpful context in the original issue that can perhaps offer ideas on next steps and/or cases to consider 🙏

akasunil committed:
reverting expand of click option changes

akasunil committed:
add change to resolve conflict

dcalhoun committed:
fix: Avoid unsupported Gallery MenuGroup usage (#63953)

The `MenuGroup` and `MenuItem` components are currently undefined for
the native mobile editor. Therefore, we cannot render them in code
shared between web and native without platform conditionals. Ideally, we
add proper support for the native platform to avoid these conditionals,
and their complexity and bundle size increase.

akasunil committed:
removing lightbox changes from edit component of gallery block

akasunil commented:
> +one minor tooltip difference. Image calls it `Link` and Gallery says `Link To`. I think both can just be called `Link`

That's good suggestion.

t-hamano commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -594,6 +617,45 @@ function GalleryEdit( props ) {
+						label={ __( 'Link To' ) }
+					>
+						{ ( { onClose } ) => (
+							<MenuGroup className="blocks-gallery__link-to-control">
```
```suggestion
							<MenuGroup>
```

Nit: This class name doesn't seem to be used anywhere.

t-hamano reviewed (approved):
@dcalhoun @akasunil

I appreciate your efforts in resolving the problem 🙇‍♂️

@artemiomorales @madhusudhand 

I think this PR is ready to ship, is it ok to merge this PR before moving forward with #56310?

akasunil committed:
Removing unneccessory class

artemiomorales reviewed (commented):
> I think this PR is ready to ship, is it ok to merge this PR before moving forward with https://github.com/WordPress/gutenberg/issues/56310?

@t-hamano Yup! I think merging this PR first or in parallel makes sense.

However, for me the issue [mentioned here](https://github.com/WordPress/gutenberg/pull/62762#pullrequestreview-2199083871) is still present, namely that the link icon in the gallery toolbar is not highlighted when a selection is made.

When an option for the gallery link is selected, the icon should appear highlighted, the same as happens with the icon in image toolbar. Here's a video demonstrating the contrast:

https://github.com/user-attachments/assets/cb67c930-e8af-465b-8e4e-77704a9dbfd4

t-hamano commented:
> the link icon in the gallery toolbar is not highlighted when a selection is made.

From my understanding, this is expected. Because this dropdown is just a tool to change the `linkDestination` attribute of Image blocks in bulk.

Indeed, the Gallery block itself has the `linkTo` attribute, but it is only used to maintain the selection state of the select box (or the dropdown menu in this PR) and is not used anywhere else.

For example, change the "Link To" value in the Gallery block to "Link images to media files". Then the user unlinks all the images in it. It seems strange that the link icon is highlighted in the block toolbar of the Gallery block, even though all the images don't have links or lightboxes.

akasunil commented:
Shall we merge this PR?

t-hamano commented:
@akasunil 

Before merging this PR, let's all agree and confirm that the issue mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/62762#pullrequestreview-2202884580) is indeed the problem that this PR should solve or not.

akasunil commented:
> +one minor tooltip difference. Image calls it `Link` and Gallery says `Link To`. I think both can just be called `Link`

@t-hamano can you share your thoughts on this?

t-hamano commented:
> +one minor tooltip difference. Image calls it "Link" and Gallery says "Link To". I think both can just be called "Link"

This suggestion makes sense and I think it's OK to make this change in this PR.

akasunil committed:
Update link control lable in gallery block

akasunil committed:
Update label in unit test

akasunil committed:
Update text in snapshot

madhusudhand reviewed (approved):
Changes looks good and Let's merge this PR. 🚀 

@akasunil A followup PR can be created [for this issue](https://github.com/WordPress/gutenberg/pull/62762#pullrequestreview-2199083871) (highlighting link icon) for design feedback and approval, based on which the change can be included or ignored.

akasunil commented:
> A followup PR can be created [for this issue](https://github.com/WordPress/gutenberg/pull/62762#pullrequestreview-2199083871) (highlighting link icon) for design feedback and approval, based on which the change can be included or ignored.

I have addressed this issue on PR https://github.com/WordPress/gutenberg/pull/64014

jeffpaul commented:
@randomburner and @RCNeil mind sharing your WordPress.org username so I can ensure its included in the 6.7 credits listing?

RCNeil commented:
@jeffpaul it is @rcneil as well.  Thanks!
